plugins/hls-hlint-plugin/src/Ide/Plugin/Hlint.hs

```diff
@@ -0,0 +1,375 @@
+          let hscExts = EnumSet.toList (extensionFlags dflags)
+          let hlintExts = mapMaybe (GhclibParserEx.readExtension . show) hscExts
+          logm $ "hlint:getIdeas:setExtensions:" ++ show hlintExts
+          return $ flags { enabledExtensions = hlintExts }
+#else
+        moduleEx _flags = do
+          mbpm <- getParsedModule nfp
+          return $ createModule <$> mbpm
+          where createModule pm = Right (createModuleEx anns modu)
+                  where anns = pm_annotations pm
+                        modu = pm_parsed_source pm
+#endif
+
+-- ---------------------------------------------------------------------
+
+data HlintUsage
+  = HlintEnabled { cmdArgs :: [String] }
+  | HlintDisabled
+  deriving Show
+
+data GetHlintSettings = GetHlintSettings
+    deriving (Eq, Show, Typeable, Generic)
+instance Hashable GetHlintSettings
+instance NFData   GetHlintSettings
+instance NFData Hint where rnf = rwhnf
+instance NFData Classify where rnf = rwhnf
+instance NFData ParseFlags where rnf = rwhnf
+instance Show Hint where show = const "<hint>"
+instance Show ParseFlags where show = const "<parseFlags>"
+instance Binary GetHlintSettings
+
+type instance RuleResult GetHlintSettings = (ParseFlags, [Classify], Hint)
+
+getHlintSettingsRule :: HlintUsage -> Rules ()
+getHlintSettingsRule usage =
+    defineNoFile $ \GetHlintSettings ->
+      liftIO $ case usage of
+          HlintEnabled cmdArgs -> argsSettings cmdArgs
+          HlintDisabled -> fail "hlint configuration unspecified"
+
+-- ---------------------------------------------------------------------
+
+codeActionProvider :: CodeActionProvider
+codeActionProvider _lf ideState plId docId _ context = Right . LSP.List . map CACodeAction <$> getCodeActions
+  where
+
+    getCodeActions = do
+        applyOne <- applyOneActions
+        diags <- getDiagnostics ideState
+        let docNfp = toNormalizedFilePath' <$> uriToFilePath' (docId ^. LSP.uri)
+            numHintsInDoc = length
+              [d | (nfp, _, d) <- diags
+                 , d ^. LSP.source == Just "hlint"
+                 , Just nfp == docNfp
+              ]
+        -- We only want to show the applyAll code action if there is more than 1
+        -- hint in the current document
+        if numHintsInDoc >= 2 then do
+          applyAll <- applyAllAction
+          pure $ applyAll:applyOne
+        else
+          pure applyOne
+
+    applyAllAction = do
+      let args = Just [toJSON (docId ^. LSP.uri)]
+      cmd <- mkLspCommand plId "applyAll" "Apply all hints" args
+      pure $ LSP.CodeAction "Apply all hints" (Just LSP.CodeActionQuickFix) Nothing Nothing (Just cmd)
+
+    applyOneActions :: IO [LSP.CodeAction]
+    applyOneActions = catMaybes <$> mapM mkHlintAction (filter validCommand diags)
+
+    -- |Some hints do not have an associated refactoring
+    validCommand (LSP.Diagnostic _ _ (Just (LSP.StringValue code)) (Just "hlint") _ _ _) =
+      code /= "Eta reduce"
```

Comment:
What's this here for? Why does eta reduction have anything to do with whether a hint has a valid refactoring? You can just ask the hint. Maybe that should go into the code or something? Or store them to the isdE?

Reply:
This was inherited from the hie plugin, so i suppose you could not ask hints for that info at that point (??)
Will investigate and change

Reply:
Ok, i've used a diagnostics code prefix to encode if a hint has refactorings as you suggested. 
It is somewhat hacky but you cant fill lsp diagnostics with custom info so :man_shrugging: .

Reply:
Is it possible that `eta reduce` has refactorings depending on ghc version?

Reply:
I think some eta reductions have hints depending on HLint version. But more broadly, there are lots of hints that sometimes don't have replacements.

Reply:
Yeah, build were using differents hlint version (3.1.4 and 3.2), i am gonna try to unify to get the same behaviour for `eta reduce`

Reply:
After using hlint-3.2 (hlint) tests are green now gor all ghc version :smiley_cat: